espnet2/asr/transducer/beam_search_transducer.py

```diff
@@ -0,0 +1,679 @@
+            max_sym_exp: number of maximum symbol expansions at each time step ("tsd")
+            u_max: maximum output sequence length ("alsd")
+            nstep: number of maximum expansion steps at each time step ("nsc")
+            prefix_alpha: maximum prefix length in prefix search ("nsc")
+            score_norm: normalize final scores by length ("default")
+        """
+        assert check_argument_types()
+
+        self.decoder = decoder
+        self.joint_network = joint_network
+        self.beam_size = beam_size
+
+        self.hidden_size = decoder.dunits
+        self.vocab_size = decoder.odim
+        self.blank = decoder.blank
+
+        if self.beam_size <= 1:
+            self.search_algorithm = self.greedy_search
+        elif search_type == "default":
+            self.search_algorithm = self.default_beam_search
+        elif search_type == "tsd":
+            self.search_algorithm = self.time_sync_decoding
+        elif search_type == "alsd":
+            self.search_algorithm = self.align_length_sync_decoding
+        elif search_type == "nsc":
+            self.search_algorithm = self.nsc_beam_search
+        else:
+            raise NotImplementedError
+
+        self.lm = lm
+        self.lm_weight = lm_weight
+
+        self.max_sym_exp = max_sym_exp
+        self.u_max = u_max
+        self.nstep = nstep
+        self.prefix_alpha = prefix_alpha
+        self.score_norm = score_norm
+
+    def __call__(self, h: torch.Tensor) -> List[Hypothesis]:
+        """Perform beam search.
+
+        Args:
+            h: Encoded speech features (T_max, D_enc)
+
+        Returns:
+            nbest_hyps: N-best decoding results
+
+        """
+        if hasattr(self.decoder, "att_list"):
+            self.decoder.att_list[0].reset()
+
+        if hasattr(self.decoder, "att"):
+            self.decoder.att[0].reset()
+
+        nbest_hyps = self.search_algorithm(h)
+
+        return nbest_hyps
+
+    def sort_nbest(self, hyps: List[Hypothesis]) -> List[Hypothesis]:
+        """Sort hypotheses by score or score given sequence length.
+
+        Args:
+            hyps: list of hypotheses
+
+        Return:
+            hyps: sorted list of hypotheses
+
+        """
+        if self.score_norm:
+            return sorted(hyps, key=lambda x: x.score / len(x.yseq), reverse=True)
+        else:
+            return sorted(hyps, key=lambda x: x.score, reverse=True)
+
+    def greedy_search(self, h: torch.Tensor) -> List[Hypothesis]:
+        """Greedy search implementation for transformer-transducer.
+
+        Args:
+            h: Encoded speech features (T_max, D_enc)
+
+        Returns:
+            hyp: 1-best decoding results
+
+        """
+        init_tensor = h.unsqueeze(0)
+        dec_state = self.decoder.init_state(init_tensor)
+
+        hyp = Hypothesis(score=0.0, yseq=[self.blank], dec_state=dec_state)
+
+        cache = {}
+
+        y, state, _ = self.decoder.score_transducer(hyp, cache, init_tensor)
```

Comment:
`score_transducer` is confusing name because it doesn't score, but return the network output.

Reply:
Yes, I used `score` and `batch_score` as request in espnet1 to try matching Scorer, I followed the same style here.
I don't have a good idea for the naming here, maybe`one_step` and `one_batch_step`? Dunno...

Reply:
`one_step` is not bad. Or just `step`?

Reply:
OK for `step_transducer` and `batch_step_transducer`!